packet/src/pacer.rs

```diff
@@ -0,0 +1,1032 @@
+        // We don't care
+    }
+
+    fn poll_timeout(&self) -> Option<Instant> {
+        if self.need_immediate_timeout {
+            self.last_sends.values().min().copied()
+        } else {
+            None
+        }
+    }
+
+    fn handle_timeout(&mut self, _now: Instant, iter: impl Iterator<Item = QueueState>) {
+        self.need_immediate_timeout = false;
+        self.queue_states.clear();
+        self.queue_states.extend(iter);
+    }
+
+    fn poll_action(&mut self) -> PollOutcome {
+        let non_empty_queues = self.queue_states.iter().filter(|q| q.packet_count > 0);
+        // Pick a queue using round robin, prioritize the least recently sent on queue.
+        let to_send_on = non_empty_queues.min_by_key(|q| self.last_sends.get(&q.id));
+
+        self.need_immediate_timeout = true;
+        to_send_on.into()
+    }
+
+    fn register_send(&mut self, now: Instant, _packet_size: DataSize, from: QueueId) {
+        let e = self.last_sends.entry(from).or_insert(now);
+        *e = now;
+    }
+}
+
+/// A leaky bucket pacer that can overshoot the target bitrate when required.
+pub struct LeakyBucketPacer {
+    /// Target bitrate.
+    target_bitrate: Bitrate,
+    /// Adjusted bitrate for when we need to drain queues.
+    adjusted_bitrate: Bitrate,
+    /// The last time we refreshed media debt and potentially adjusted the bitrate.
+    last_handle_time: Option<Instant>,
+    /// The last time we indicated that a packet should be sent.
+    last_send_time: Option<Instant>,
+    /// The next time we should send a queued packet.
+    next_send_time: Option<Instant>,
+    /// The current media debt.
+    media_debt: DataSize,
+    /// The current pacing i.e. how frequently we clear out debt and when we are exceeding the
+    /// target bitrate how long we wait to send.
+    pacing: Duration,
+    /// The longest the average packet can spend in the queue before we force it to be drained.
+    queue_limit: Duration,
+    /// The queue states given by last handle_timeout.
+    queue_states: Vec<QueueState>,
+
+    need_immediate_timeout: bool,
+
+    next_poll_outcome: Option<PollOutcome>,
+}
+
+impl Pacer for LeakyBucketPacer {
+    fn set_target_bitrate(&mut self, new_rate: Bitrate) {
+        self.target_bitrate = new_rate;
+    }
+
+    fn poll_timeout(&self) -> Option<Instant> {
+        if self.need_immediate_timeout {
+            return self.last_send_time;
+        }
+
+        let next_handle_time = self.last_handle_time.map(|lh| lh + self.pacing);
+
+        match (next_handle_time, self.next_send_time) {
+            (Some(nh), Some(ns)) => Some(nh.min(ns)),
+            (None, Some(ls)) => Some(ls),
+            (Some(nh), None) => Some(nh),
+            (None, None) => None,
+        }
+    }
+
+    fn handle_timeout(&mut self, now: Instant, iter: impl Iterator<Item = QueueState>) {
+        self.need_immediate_timeout = false;
+        self.queue_states.clear();
+        self.queue_states.extend(iter);
```

Comment:
Is this not the same as self.queue_states = iter.collect()?  If so, why not just always pass in a Vec<QueueState>?